pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java

```diff
@@ -190,55 +181,32 @@ public boolean readCompacted() {
      *
      * @return a SendMessageInfo object that contains the detail of what was sent to consumer
      */
-    public SendMessageInfo sendMessages(final List<Entry> entries) {
-        // Empty listener
-        return sendMessages(entries, null);
-    }
-
-    /**
-     * Dispatch a list of entries to the consumer. <br/>
-     * <b>It is also responsible to release entries data and recycle entries object.</b>
-     *
-     * @return a SendMessageInfo object that contains the detail of what was sent to consumer
-     */
-    public SendMessageInfo sendMessages(final List<Entry> entries, SendListener listener) {
+    public ChannelPromise sendMessages(final List<Entry> entries, int[] batchSizes, SendMessageInfo sendMessageInfo) {
         final ChannelHandlerContext ctx = cnx.ctx();
-        final SendMessageInfo sentMessages = new SendMessageInfo();
-        final ChannelPromise writePromise = listener != null ? ctx.newPromise() : ctx.voidPromise();
-
-        if (listener != null) {
-            writePromise.addListener(future -> listener.sendComplete(writePromise, sentMessages));
-        }
+        final ChannelPromise writePromise = ctx.newPromise();
 
-        if (entries.isEmpty()) {
+        if (entries.isEmpty() || sendMessageInfo.getTotalMessages() == 0) {
             if (log.isDebugEnabled()) {
                 log.debug("[{}-{}] List of messages is empty, triggering write future immediately for consumerId {}",
                         topicName, subscription, consumerId);
             }
             writePromise.setSuccess();
-            sentMessages.totalSentMessages = 0;
-            sentMessages.totalSentMessageBytes = 0;
-            return sentMessages;
+            return writePromise;
         }
 
-        try {
-            updatePermitsAndPendingAcks(entries, sentMessages);
-        } catch (PulsarServerException pe) {
-            log.warn("[{}] [{}] consumer doesn't support batch-message {}", subscription, consumerId,
-                    cnx.getRemoteEndpointProtocolVersion());
-
-            subscription.markTopicWithBatchMessagePublished();
-            sentMessages.totalSentMessages = 0;
-            sentMessages.totalSentMessageBytes = 0;
-            // disconnect consumer: it will update dispatcher's availablePermits and resend pendingAck-messages of this
-            // consumer to other consumer
-            disconnect();
-            return sentMessages;
-        }
+        // reduce permit and increment unackedMsg count with total number of messages in batch-msgs
+        MESSAGE_PERMITS_UPDATER.addAndGet(this, -sendMessageInfo.getTotalMessages());
+        incrementUnackedMessages(sendMessageInfo.getTotalMessages());
+        msgOut.recordMultipleEvents(sendMessageInfo.getTotalMessages(), sendMessageInfo.getTotalBytes());
 
         ctx.channel().eventLoop().execute(() -> {
```

Comment:
I would move this call to another method which is only given the entries, to avoid the temptation to touch batchSizes or the message Info in the callback.

Reply:
Good point. there was actually a problem there already. Since I need to check the sizes from the loop anyway, I just changed it to use a recyclable instead of the thread local. Then it becomes easier and cleaner.